### PR TITLE
Version Model Update for Pull Request Builder

### PR DESCRIPTION
**Pull request by saadmk11**
This PR will Update the Version Model to be used for Pull Request Builder.

### **Tasks:**

- [x] Add pull request `VERSION_TYPES`
- [x] Update `Version` Model Methods
- [x] Add `Version` Model `Manager` for `internal` and `external` Versions
- [x] Add tests for `Querysets`

We need to think about how we want to create slugs for PR Versions. Should the slug be like this
`pr-<number>` or only `<pr_number>`. We have to think about how we can make pr version slugs unique so that it doesn't cause problems for the normal Versions. if we do `pr-<number>`  the `vcs_url()` method of the `Version` Model needs to be updated.

Feel free to let me know if there are other places that we need to update for this PR.

I would like to get some initial feedback's to continue the work on slug creation and writing tests.

### CODE REVIEW

**Comment by stsewd** [comment]
We may want to use managers or another pattern, feels like a lot of places where we can forget to exclude PRs from the actual versions

File: readthedocs/api/v2/views/model_views.py [hunk 1]
```diff
@@ -130,7 +130,7 @@ def active_versions(self, request, **kwargs):
             Project.objects.api(request.user),
             pk=kwargs['pk'],
         )
-        versions = project.versions.filter(active=True)
+        versions = project.versions(manager='internal').filter(active=True)
```
Comment (ericholscher) [inline readthedocs/api/v2/views/model_views.py]: Is this the normal way to handle this? Is it not normally `project.versions.internal.filter`?

File: readthedocs/core/management/commands/update_repos.py [hunk 2]
```diff
@@ -71,6 +71,48 @@ def handle(self, *args, **options):
                             state='triggered',
                         )
 
+                        # pylint: disable=no-value-for-parameter
+                        tasks.update_docs_task(
+                            version.project_id,
+                            build_pk=build.pk,
+                            version_pk=version.pk,
+                        )
+                elif version == 'internal':
```
Comment (ericholscher) [inline readthedocs/core/management/commands/update_repos.py]: These options should be documented somewhere. Also we should be using constants for these, not strings `INTERNAL = 'internal'` in `projects/constants.py` or similar.

File: readthedocs/api/v2/views/model_views.py (hunk 1 above)
Comment (saadmk11) [reply readthedocs/api/v2/views/model_views.py]: yes this is the normal way to do this if we use multiple managers.
ref: [Dajngo Docs](https://docs.djangoproject.com/en/1.11/topics/db/queries/#using-a-custom-reverse-manager)

File: readthedocs/core/management/commands/update_repos.py (hunk 2 above)
Comment (saadmk11) [reply readthedocs/core/management/commands/update_repos.py]: Okay I'll Add them in the constants. Do we Document Management Commands Somewhere?

Comment (ericholscher) [reply readthedocs/core/management/commands/update_repos.py]: Not currently -- perhaps we should do something like this: https://docs.readthedocs.io/en/latest/guides/feature-flags.html

Comment (saadmk11) [reply readthedocs/core/management/commands/update_repos.py]: Okay i'll Add these in another PR.

File: readthedocs/builds/managers.py [hunk 3]
```diff
@@ -85,6 +86,41 @@ def get_object_or_log(self, **kwargs):
...
+
+
 class VersionManager(SettingsOverrideObject):
     _default_class = VersionManagerBase
     _override_setting = 'VERSION_MANAGER'
+
+
+class InternalVersionManager(SettingsOverrideObject):
+    _default_class = InternalVersionManagerBase
+    _override_setting = 'INTERNAL_VERSION_MANAGER'
```
Comment (ericholscher) [inline readthedocs/builds/managers.py]: Let's remove these `_override_setting` variable from all these managers. We have another way to override them explicitly now.

Comment (saadmk11) [reply readthedocs/builds/managers.py]: Done!

File: readthedocs/builds/constants.py [hunk 4]
```diff
@@ -33,11 +33,13 @@
 
 BRANCH = 'branch'
 TAG = 'tag'
+PULL_REQUEST = 'pull_request'
```
Comment (humitos) [inline readthedocs/builds/constants.py]: Shouldn't this be called `external` to keep consistency with the rest of the code (internal/external)?

**Comment by ericholscher** [comment]
> we agreed on using internal/external, but there are some places where we call the Versions as Pull Requests. I think it would be good to have all using the same term for consistency.

Agreed. :+1:

> I remember that we talked about maybe override .objects so we raise a Warning or fail when it's used to force the developer to use either .internal or .external. Should that be included in this PR as well?

We have that in #5806 it needs more work.

File: readthedocs/builds/constants.py (hunk 4 above)
Comment (saadmk11) [reply readthedocs/builds/constants.py]: Thanks for the review i'll update everything to `internal` and `external` :)

**Comment by saadmk11** [comment]
@humitos @ericholscher  Updated the PR :)

**Comment by stsewd** [comment]
We should check that `Version.internal.public` and `Version.external.public` is doing the filter correctly, because `public` can add all the versions ignoring any previous filter (see #5783)

**Comment by stsewd** [comment]
About the slug, doesn't matter what we choose, the slug is unique. We add prefixes to make a slug unique if there is one with that slug.

User creates a branch `pr-200`, the slug is created as `pr-200`, if a pull requests #200 is created the slug is `pr-200_a`.

**Comment by saadmk11** [comment]
> We should check that `Version.internal.public` and `Version.external.public` is doing the filter correctly, because `public` can add all the versions ignoring any previous filter (see #5783)

I have already added some tests. I'll add some more to be more careful.

> About the slug, doesn't matter what we choose, the slug is unique. We add prefixes to make a slug unique if there is one with that slug.

> User creates a branch pr-200, the slug is created as pr-200, if a pull requests #200 is created the slug is pr-200_a.

Then we have to add the prefix (`pr-`) for the External version slugs. how do you think we should handle `vcs_url()` ?

**Comment by stsewd** [comment]
In the version model we have `identifier` and `verbose_name`, I guess we are going to add the commit hash in identifier, and we can add the pr indentifier in `verbose_name`, this should be `123` (not pr-123) for github or anything that is used by the provider to identify a pr. That way is more easy to build the url

**Comment by ericholscher** [comment]
Closing this in favor of #5823, which should now be our feature branch for this work until it's merged into master.

**Pull request closed without merging**